File: codeiq/codeiq1700.cpp
#include <map>
std::map<long int, int> memo;
int n = 22;

int solve(int used, int status)
{
	long int hash_key = used << n | status;
	auto itr = memo.find(hash_key);
	if (itr != memo.end())
	{
		return memo[hash_key];
	}
	if (used == n / 2)
	{
		return 1;
	}
	int num = used + 1;
	int space = num + 1;
	int result = 0;
	int l, r;
	for (int i = 0; i < n - space; i++)
	{
		l = 1 << i;
		r = l << space;
		if (((status & l) == 0) && ((status & r) == 0))
		{
			result += solve(num, status | l | r);
		}
	}
	memo[hash_key] = result;
	return result;
}

int main(int argc, char* argv[])
{
	int res = solve(0, 0);
	printf("%d\n", res);
}

File: codeiq/codeiq1700.rb
@memo = {}
@n = 22

def solve(used, status)
	p "for used:#{used.to_s(2)}, status:#{status.to_s(2)}"
	mem = @memo[(used << @n) | status]
	return mem if mem
	return 1 if used == @n / 2
	num = used + 1
	result = 0
	(@n - num - 1).times { |i|
		l = 1 << i
		r = 1 << i + num + 1
		if status & l == 0 && status & r == 0
			result += solve(num, status | l | r)
		end
	}
	@memo[(used << @n) | status] = result
	return result
end

used = 0
status = 0

puts solve(used, status)